src/amber/router/pipe/session.cr

```diff
@@ -3,48 +3,77 @@ require "base64"
+    # encode and decode the cookie and provide the hash in the context that can
+    # be used to maintain data across requests.
+    class Session < Base
+      property :key
+      property secret : String
 
-    def initialize
-      @key = "#{Server.settings.project_name}.session"
-      @secret = Server.settings.secret
-    end
+      def self.instance
+        @@instance ||= new
+      end
 
-    def call(context : HTTP::Server::Context)
-      cookies = context.request.cookies
-      decode(context.session, cookies[@key].value) if cookies.has_key?(@key)
-      call_next(context)
-      value = encode(context.session)
-      cookies = context.response.cookies
-      cookies << HTTP::Cookie.new(@key, value)
-      cookies.add_response_headers(context.response.headers)
-      context
-    end
+      def initialize
+        @key = "#{Server.settings.project_name}.session"
+        @secret = Server.settings.secret
+      end
 
-    private def decode(session, data)
-      sha1, data = data.split("--", 2)
-      if sha1 == OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
-        values = YAML.parse(Base64.decode_string(data))
-        values.each do |key, value|
-          session[key.to_s] = value.to_s
+      def call(context : HTTP::Server::Context)
+        cookies = context.request.cookies
+        decode(context.session, cookies[@key].value) if cookies.has_key?(@key)
+        call_next(context)
+        value = encode(context.session)
+        cookies = context.response.cookies
+        cookies << HTTP::Cookie.new(@key, value)
+        cookies.add_response_headers(context.response.headers)
+        context
+      end
+
+      private def decode(session, data)
+        sha1, data = data.split("--", 2)
+        if sha1 == OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
+          values = YAML.parse(Base64.decode_string(data))
+          values.each do |key, value|
+            session[key.to_s] = value.to_s
+          end
         end
       end
+
+      private def encode(session)
+        data = Base64.encode(session.to_yaml)
+        sha1 = OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
+        "#{sha1}--#{data}"
+      end
     end
+  end
+
+  module Router
+    module Session
+      # clear the session.  You can call this to logout a user.
+      def clear_session
+        @session = Hash.new 
+      end
+
+      # Holds a hash of session variables.  This can be used to hold data between
+      # sessions.  It's recommended to avoid holding any private data in the
+      # session since this is held in a cookie.  Also avoid putting more than 4k
+      # worth of data in the session to avoid slow pageload times.
+      def session
+        @session ||= Hash.new 
+      end
 
-    private def encode(session)
-      data = Base64.encode(session.to_yaml)
-      sha1 = OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
-      "#{sha1}--#{data}"
+      class Hash < Hash(String, String)
```

Comment:
Rename this to SessionHash to stay consistent with FlashHash. Beware that this is overwriting the crystal lib Hash class.

What do you think of this approach? 

```crystal 
module Amber
  module Pipe
     module SessionHelper
        def clear_session
          @session = Hash.new 
        end
        ...
     end

     class Session < Base
     end
  end
end
```

Then in the Context  `include Pipe::Session::Helpers`. This is a style convention that we all must agree upon so we stay consistent.